### ISSUE TITLE
Punctuation and Long Conversations

### ISSUE THREAD

**Issue by CelloCodez** [ProgrammerDan/learning-chatbot#1]
I have noticed that sometimes it messes up punctuation, and
that the longer you talk to it, the wackier its grammar gets.
I do not know if you are still working on this, but could you
link me to a webpage for an example or about the algorithm(s)
you used?

This AI Chatbot is very cool indeed, and I have gotten it to say
really cool things. I actually got it to read a text file (treating
each line or sentence in the file like input from the user) that is
formatted for each line a sentence. When I gave it The Outsider
(~84 Sentences) by H.P. Lovecraft, it talked with pretty good grammar.
But I tried feeding it some longer texts (past hundreds of lines) and,
after about 20 seconds, it would read it, then I could talk to it, but its
grammar was crazy. Could you give me any tips on modifications or ideas for
things to improve to fix this?

TL;DR This AI is the best I have ever seen open source, but you don't
seem to develop it anymore. How could I try to modify it?

**Comment by ProgrammerDan** [comment]
Thanks for the comment, and I'm excited to see someone interested in this! 

I knew from the start that after extremely long conversations (100+ interactions) you'll start to see degrading grammar. This is due to the very same algorithm that gives it success; it prefers to chain words together based on frequency of observing them together. As its sample space increases, it can chain more and more words together, leading to longer, wackier phrases. If Lovecraft uses shorter sentences and more unique combinations, the resulting grammar from the bot will appear to be "good"; longer texts will introduce a larger variety of overlapping combinations, allowing the algorithm to produce seriously confusing spates of grammar.

Out of curiosity, how'd you find the bot? Through CodeGolf?

Anyway, at the very bottom of the original post to CodeGolf: http://codegolf.stackexchange.com/questions/20914/who-is-this-chatbot/24788#24788

I give a few thoughts on where I was going to head. The biggest one is this nugget from natural language parsing:

The most frequently observed words are of the least value. For instance, "the", "of", "an", and "a" will dominate longer interactions, but they will also lead to the greatest variety of second-order connections -- and the greatest confusion in resulting grammar. There should be a decay in preference, so that as words begin to dominate an occurrence list, the bot begins to prefer not to use them in connecting phrases. They should always be available for use, but simply less likely.

One I didn't mention in the post is third-order linking. This bot uses just second-order observed linking to generate its replies; adding in a third-order match, while significantly increasing the data structure complexity, would significantly improve the results (and may invalidate the need for my prior suggestion! I'm not sure ...). It would also prettymuch require the use of a database, as you'll run out of memory to hold the structures for any significant conversation with third order linking.

As for linking you to articles or examples, I'd have to just google around myself. This technique borrows heavily from some very standard techniques in the domain of Natural Language Parsing, so you might find some relevant information using those keywords. 

I haven't kept up on it, because I wrote it on a whim for the codegulf challenge and after winning so soundly, lost a lot of motivation :). Feel free to fork the code and make pull requests, however! Although I don't make as many daily contributions to my github projects as I'd like, I'm still very active in watching activity and interest. 

Thanks for the comment, and let me know how else I can help. I look forward to seeing what you make of this!

**Comment by CelloCodez** [comment]
Thanks for all the info!

Yes, I originally saw this on CodeGolf. I sent the code to a friend,
y'know, "Hey, look at this cool Chatbot I found on the internet,"
and he is into coding a little, so it amazed him alot that it could rearrange
words and change up sentences. This truly is great!

Anyway, into ordering: what if I could change the ordering and frequency part and
start adding functions to order orders? Basically, order the orders so that more
advanced phrases can be created ("What is this" could be an example phrase prefixing another word). That is one idea. I am looking into the code, slowly understanding parts of it.
Also, you did a very good job with commenting the code. Can't wait to see what I
can figure out! I am just waiting for it to come up with its own jokes or stories.
So far it does say crazy stuff, like one time I got bored and started telling it "I'm
Batman," (Yes, I do have serious discussions with it. This was just for fun) and
"The legend of Link," (a joke because Link is the main character) and it said
something like "No one can be Batman unless you can make perfect sense of the Legend of Link"
and I started laughing. Then, it started to freak me out when I talked to it about its
universe and mine, and the whole "I think therefore I am," thing, when it basically said
something meaning that its universe only existed by my input. It can be really smart!
I have had some great conversations with this, and it has even made me question intelligence
with its whole thing about thinking and only existing every time I talk to it.

Technically, though, this AI only thinks during the period after your input, and before it outputs the sentence.

**Comment by CelloCodez** [comment]
Better to say, only 'thinks' not only 'exists' at that last sentence.

**Comment by CelloCodez** [comment]
Maybe it could recognize relationships between phrases and words, adding another level. It could recognize things like <new word not commonly used> + <common word like "means"> + <phrase> and try to get some topics from the phrase, then try to get the main topic for that word. If this thing get figure out synonymous vocabulary this way, then that would be a great start to giving it logic.

**Comment by ProgrammerDan** [comment]
Any of your suggestions would alter the outputs, absolutely! And that's the great thing about using this bot as a base for your experiments; the bot I wrote is as close to a "zero-knowledge predicate" bot as I could imagine, meaning that before you start talking to it, it knows absolutely nothing about how to converse with you. Everything it knows, it learns from you.

The only exception I had to make wes a little understanding that punctuation wasn't a part of words. That's why you'll see commas and periods "randomly" applied throughout constructed sentences after a while. I deal with it as a separate "link", and there's a probability that punctuation will appear after words you've used punctuation after. It still doesn't know that it's punctuation, it's just a symbol that you've used and it's cataloged. It won't introduce punctuation after words you have never used it after; it's still just copying you :).

Adding deeper parsing; phrase parsing, explicit topic parsing, etc -- now you're headed into predicate knowledge bots -- "traditional" chat bots -- that attempt to do some natural language parsing and semantic analysis before replying. This is _also_ a lot of fun stuff! It's totally different from the initial goals of this bot, but I see no reason you couldn't layer those concepts in successfully, as you explore and build.

If you do go this route, definitely fork this codebase and start hacking. I'd love to see where you take it, so if you do I'll be sure to follow it!

**Comment by CelloCodez** [comment]
Ok! I also had an idea of treating punctuation similar to words, like a grammar type of word, that the only knowledge you would have to program into the bot is to have no space between a word and the following punctuation.

**Comment by ProgrammerDan** [comment]
That is _similar_ to the current behavior, although I'm sure it could be improved!

**Comment by ProgrammerDan** [comment]
Hey, curiosity strikes -- have you taken any time to work on your own modification to this chatbot? If you have, I'd love to see/hear about what you've done.

**Comment by CelloCodez** [comment]
Sorry and sadly, but no. It was more of inspirations to work on this, but I have had other interests and not enough time to put aside all that I want to do. But if I ever do get a chance, your chatbot is still the best one that I can get the source for on the internet! Really, I mean it though. I have had some cool (and awkward) moments with this chatbot; it even one time saying something awesome that made scientific sense!

**Comment by ProgrammerDan** [comment]
Sounds awesome :D. Be sure to let me know if you do work on it, I'd love to see it. For now, I'll close this issue.
